src/read/reducer.js

```diff
@@ -19,11 +19,31 @@ import { ReadFn } from './readRegistry';
  */
 export default function(cursor, action) {
   // console.log('read action', action);
-  const canonicalKind = canonical(cursor.getIn(action.fromPath).get('kind'));
+  const element = cursor.getIn(action.fromPath);
+  if(!element) {
+    // the path for the action can't be accessed in the latest cursor
+    // cursor has changed, so we can only discard the action
+    return cursor;
+  }
+  const canonicalKind = canonical(element.get('kind'));
   const pathToKind = List.of(...action.fromPath, 'kind');
+  if (action.random) {
+    const pathToRandom = List.of(...action.fromPath, 'random');
+    if (action.random !== cursor.getIn(pathToRandom)) {
```

Comment:
This should happen only in the READ_SUCCEEDED / READ_FAILED branches of the reducer.